File: frontend/src/types/document.ts
export type DocumentStatus =
  | "uploaded"
  | "processing"
  | "completed"
  | "indexed"
  | "failed"

export interface KnowledgeDocument {
  id: string
  user_id: number

  title: string
  file_name: string
  original_file_name?: string | null
  file_path?: string | null

  document_type: string
  department?: string | null
  description?: string | null

  file_size: number
  status: DocumentStatus | string

  processing_progress: number
  chunk_count: number

  embedding_model?: string | null
  vector_dimension?: number | null
  vector_collection?: string | null

  version_number?: number
  is_deleted?: boolean

  created_at: string
  updated_at?: string
}

export interface DocumentListResponse {
  items: KnowledgeDocument[]
  total: number
  page: number
  page_size: number
  total_pages?: number
}

export interface DocumentUploadResponse {
  message: string
  document: KnowledgeDocument
}

export interface DocumentDeleteResponse {
  message: string
  document_id: string
}

export interface DocumentIndexResponse {
  message: string
  document_id: string
  status: string
  chunk_count?: number
}

export interface DocumentListParams {
  page?: number
  page_size?: number
  search?: string
  department?: string
  status?: string
  document_type?: string
}

export interface UploadDocumentPayload {
  file: File
  title?: string
  department?: string
  document_type?: string
  description?: string
}

File: backend/app/schemas/document.py
from __future__ import annotations

from datetime import datetime
from uuid import UUID

from pydantic import (
    BaseModel,
    ConfigDict,
    Field,
)


class DocumentBase(BaseModel):
    title: str = Field(
        ...,
        min_length=1,
        max_length=255,
    )

    document_type: str = Field(
        ...,
        min_length=1,
        max_length=50,
    )

    department: str | None = Field(
        default=None,
        max_length=100,
    )

    description: str | None = None


class DocumentCreate(DocumentBase):
    """
    Metadata received together with the uploaded file.
    """

    pass


class DocumentUpdate(BaseModel):
    title: str | None = Field(
        default=None,
        min_length=1,
        max_length=255,
    )

    document_type: str | None = Field(
        default=None,
        min_length=1,
        max_length=50,
    )

    department: str | None = Field(
        default=None,
        max_length=100,
    )

    description: str | None = None


class DocumentResponse(BaseModel):
    model_config = ConfigDict(
        from_attributes=True,
    )

    id: UUID
    user_id: int

    title: str
    file_name: str
    original_file_name: str | None

    document_type: str
    department: str | None
    description: str | None

    file_size: int
    file_path: str
    file_hash: str | None

    status: str
    version_number: int
    processing_progress: int
    chunk_count: int

    embedding_model: str | None
    vector_dimension: int | None
    vector_collection: str | None

    is_deleted: bool

    created_at: datetime
    updated_at: datetime
    deleted_at: datetime | None


class DocumentUploadResponse(BaseModel):
    message: str
    document: DocumentResponse


class DocumentListResponse(BaseModel):
    items: list[DocumentResponse]
    total: int
    page: int
    page_size: int
    total_pages: int


class DocumentStatusResponse(BaseModel):
    id: UUID
    status: str
    processing_progress: int
    chunk_count: int
    embedding_model: str | None
    vector_dimension: int | None
    vector_collection: str | None


class DocumentDeleteResponse(BaseModel):
    message: str
    document_id: UUID
